runtime/warp.cu

```diff
@@ -248,4 +248,50 @@ __device__ void iterGroupedStaticWarpAllReduce(
   // reduction is done.
   block_sync::sync<Aligned>(block_dim, barrier_id);
 }
+
+template <
+    bool SINGLE_WARP,
+    bool Aligned,
+    int n_threads,
+    typename T,
+    typename Func>
+__device__ void staticWarpAllReduceTIDX(
+    T& out,
+    const T& inp_val,
+    Func reduction_op,
+    T* shared_mem,
+    uint32_t barrier_id = 1) {
+  constexpr int WARP_SIZE = 32;
+  constexpr dim3 block_dim = dim3(n_threads, 1, 1);
+  constexpr unsigned int num_of_warps = n_threads / WARP_SIZE;
+  T reduce_val = inp_val;
+  // Reduce within each warp
+  for (int i = 16; i >= 1; i /= 2) {
+    reduction_op(reduce_val, shfl_xor(reduce_val, i, WARP_SIZE));
+  }
+
+  // Reduce across warp if needed
+  // Load value to shared mem
+  if constexpr (!SINGLE_WARP) {
+    unsigned int warp_idx = threadIdx.x / WARP_SIZE;
+    unsigned int lane_idx = threadIdx.x % WARP_SIZE;
+
+    if (lane_idx == 0) {
+      shared_mem[warp_idx] = reduce_val;
+    }
+    block_sync::sync<Aligned>(block_dim, barrier_id);
+
+    // All reduce
+    out = shared_mem[0];
```

Comment:
I'm probably missing something obvious, shouldn't we predicate this on `if (lane_idx == 0)`?